src/dupin.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <Eigen/Dense>
+#include <algorithm>
+#include <iostream>
+#include <limits>
+#include <omp.h>
+#include <unordered_map>
+#include <vector>
+
+// DynamicProgramming class for dynamic programming based segmentation.
+class DynamicProgramming { // change name to Dynamic Programming
+private:
+  class upper_triangular_cost_matrix {
+  private:
+    std::vector<double> matrix;
+    int size;
+
+    int index(int i, int j) const {
+      return i * (2 * size - i + 1) / 2 + (j - i);
+    }
+

```

Comment:
```suggestion
    int index(int col, int row) const {
      return (col * (2 * size - col + 1) / 2) - col + row;
    }


```

Reply:
Oh wait I think the row and column is swapped, I'll fix in the next commit